privval/signer_listener_endpoint.go

```diff
@@ -0,0 +1,227 @@
+	sl.mtx.Lock()
+	defer sl.mtx.Unlock()
+	return sl.ensureConnection(maxWait)
+}
+
+// SendRequest ensures there is a connection, sends a request and waits for a response
+func (sl *SignerListenerEndpoint) SendRequest(request RemoteSignerMsg) (RemoteSignerMsg, error) {
+	sl.mtx.Lock()
+	defer sl.mtx.Unlock()
+
+	sl.Logger.Debug("SignerListener::EnsureConnection")
+	err := sl.ensureConnection(sl.timeoutAccept)
+	if err != nil {
+		return nil, err
+	}
+
+	sl.Logger.Debug("SignerListener::Write")
+	err = sl.WriteMessage(request)
+	if err != nil {
+		sl.Logger.Debug("SignerListener::Write error")
+		return nil, err
+	}
+	sl.Logger.Debug("SignerListener::WriteFinished")
+
+	sl.Logger.Debug("SignerListener::Read")
+	res, err := sl.ReadMessage()
+	if err != nil {
+		return nil, err
+	}
+
+	return res, nil
+}
+
+func (sl *SignerListenerEndpoint) ensureConnection(maxWait time.Duration) error {
+	if !sl.isConnected() {
+		// Is there a connection ready?
+		select {
+		case sl.conn = <-sl.connectedCh:
+			return nil
+		default:
+		}
+
+		// should we trigger a reconnect?
+		select {
+		case sl.connectCh <- struct{}{}:
+			break
+		default:
+			break
+		}
+
+		// block until connected or timeout
+		select {
+		case sl.conn = <-sl.connectedCh:
+			break
+		case <-time.After(maxWait):
+			return ErrListenerTimeout
+		}
+	}
+
+	return nil
+}
+
+func (sl *SignerListenerEndpoint) acceptNewConnection() (net.Conn, error) {
+	if !sl.IsRunning() || sl.listener == nil {
+		return nil, fmt.Errorf("endpoint is closing")
+	}
+
+	// wait for a new conn
+	conn, err := sl.listener.Accept()
+	if err != nil {
+		return nil, err
+	}
+
+	sl.Logger.Info("SignerListener: New connection")
+	return conn, nil
+}
+
+func (sl *SignerListenerEndpoint) serviceLoop() {
+	defer close(sl.stoppedServiceLoopCh)
+
+	for {
+		select {
+		case <-sl.connectCh:
+			{
+				sl.Logger.Info("SignerListener: Listening for new connection")
+
+				conn, err := sl.acceptNewConnection()
+				if err == nil {
+					sl.Logger.Info("SignerListener: Connected")
+
+					// We have a good connection, wait for someone that needs one or cancellation
+					select {
+					case sl.connectedCh <- conn:
+						sl.Logger.Debug("SignerListener: Returned connection")
+					case <-sl.stopServiceLoopCh:
+						sl.Logger.Debug("SignerListener: Stopping service")
+						return
+					}
+				}
+
+				select {
+				case sl.connectCh <- struct{}{}:
+				default:
+				}
+
+				sl.Logger.Debug("SignerListener: service loop end ")
+			}
+		case <-sl.stopServiceLoopCh:
+			return
+		}
+	}
+}
+
+func (sl *SignerListenerEndpoint) pingLoop() {
+	defer close(sl.stoppedPingLoopCh)
+
+	for {
+		select {
+		case <-sl.pingTimer.C:
+			{
+				sl.Logger.Info("Ping!")
+				_, err := sl.SendRequest(&PingRequest{})
+
+				if err != nil {
+					sl.Logger.Error("SignerListener: Ping timeout")
+					// Drop and try to reconnect
+					sl.DropConnection()
+					sl.connectCh <- struct{}{}
```

Comment:
I find the whole logic with multiple connect and close channels is a bit hard to follow. There maybe bugs / deadlocks here. I guess we'll see

Reply:
It is actually not too complicated. 
I will refactor it a bit more so it is easier to follow.